### PR TITLE
Add workerCount option

### LINKED ISSUE #2022: Add an option to set the number of spawned WebWorkers
I am working on the app where we need to be able to control / reduce the number of web workers that mapbox-gl uses. Looks like now it is always _navigator.hardwareConcurrency - 1_ ?  Is there a reason why the number of workers should not be exposed / overwritten via API?

### PR DESCRIPTION
**Pull request by anandthakker**
Replaces #2655

Also sets workerCount on the debug page to 3, on the theory that a pretty large majority of users have 4 cpus.

Closes #2022

### CODE REVIEW

File: debug/index.html [hunk 1]
```diff
@@ -58,7 +58,8 @@
     zoom: 12.5,
     center: [-77.01866, 38.888],
     style: 'mapbox://styles/mapbox/streets-v8',
-    hash: true
+    hash: true,
+    workerCount: 3
```
Comment (mourner) [inline debug/index.html]: I don't think we need it set in the debug page — let's leave it to rely on `hardwareConcurrency`, which most of the browsers used by GL JS developers support. Or we could add but comment out.

Comment (anandthakker) [inline debug/index.html]: @mourner My thinking was that, if the majority of users have 4 cores, it might actually be beneficial to have developers working under those conditions by default.

Comment (lucaswoj) [inline debug/index.html]: I've been doing a little housecleaning on this debug page to keep it as simple as possible. My intent (we're not there yet) is for this page to be as simple as possible, making it an ideal place to build one-off testing rigs, and to put any persistent specialized setups on other pages (i.e. `video.html`)

tl;dr 👍 @mourner

Comment (anandthakker) [inline debug/index.html]: 👍

**Comment by anandthakker** [comment]
@lucaswoj @mourner amended the commit to remove the change to debug/index.html

File: js/ui/map.js [hunk 2]
```diff
@@ -97,6 +98,7 @@ var defaultOptions = {
  * @param {number} [options.zoom] The zoom level of the map's initial viewport.
  * @param {number} [options.bearing] The bearing (rotation) of the map's initial viewport measured in degrees counter-clockwise from north.
  * @param {number} [options.pitch] The pitch of the map's initial viewport measured in degrees.
+ * @param {number} [options.workerCount=navigator.hardwareConcurrency] The number of WebWorkers the map should use to process vector tile data.
```
Comment (lucaswoj) [inline js/ui/map.js]: `options.workerCount=navigator.hardwareConcurrency `**`- 1`**

File: js/ui/map.js [hunk 3]
```diff
@@ -586,7 +594,7 @@ util.extend(Map.prototype, /** @lends Map.prototype */{
         } else if (style instanceof Style) {
             this.style = style;
         } else {
-            this.style = new Style(style, this.animationLoop);
+            this.style = new Style(style, this.animationLoop, this._workerCount);
```
Comment (lucaswoj) [inline js/ui/map.js]: I have a preference towards passing `options.workerCount` directly to `Style`. I'm trying to push `Map` in the direction of becoming a dumb [facade](https://en.wikipedia.org/wiki/Facade_pattern) class because it is currently a [god object](https://en.wikipedia.org/wiki/God_object).

File: test/js/ui/map.test.js [hunk 4]
```diff
@@ -930,6 +931,21 @@ test('Map', function(t) {
         t.end();
     });
 
+    t.test('workerCount option', function(t) {
+        var map = createMap({});
+        // TODO: it's not great to check this private member here; better would
+        // be to mock Style and then just check that the correct workerCount
+        // param is being passed.
```
Comment (lucaswoj) [inline test/js/ui/map.test.js]: We have a policy against checking `// TODO` comments into `master` because they tend to get forgotten. You should either resolve the todo or open a ticket.

File: test/js/ui/map.test.js [hunk 5]
```diff
@@ -930,6 +931,21 @@ test('Map', function(t) {
...
+        t.equal(map._workerCount, browser.hardwareConcurrency - 1, 'workerCount defaults to hardwareConcurrency - 1');
+        map = createMap({ workerCount: 3 });
+        t.equal(map._workerCount, 3, 'workerCount option is used');
```
Comment (lucaswoj) [inline test/js/ui/map.test.js]: How 'bout checking `map.style.dispatcher.actors.length` here? It also accesses private properties but is 🆗 for now.

**Comment by lucaswoj** [comment]
Feel free to 🚢 once you've addressed my comments above!

File: js/ui/map.js (hunk 3 above)
Comment (anandthakker) [inline js/ui/map.js]: @lucaswoj As long as `Map` has a `setStyle` method, I think it will need to keep a copy like this of the `workerCount`, because it needs to be able to do `new Style(...)` not just in the map's constructor.  We _could_ add `workerCount` as a parameter to `setStyle`, but that seems a little odd to me.

File: test/js/ui/map.test.js (hunk 4 above)
Comment (anandthakker) [inline test/js/ui/map.test.js]: Ticketed here: https://github.com/mapbox/mapbox-gl-js/issues/2685

**Comment by anandthakker** [comment]
@lucaswoj most comments addressed, but see my note [above](https://github.com/mapbox/mapbox-gl-js/pull/2666#discussion_r66053368)

File: js/ui/map.js (hunk 3 above)
Comment (lucaswoj) [inline js/ui/map.js]: Ah! I didn't consider the `setStyle` case. I agree that this `this._workerCount` is necessary for now 👍

**Comment by lucaswoj** [comment]
Looks good! :shipit:
